src/main/java/org/apache/mesos/executor/HealthCheckHandler.java

```diff
@@ -0,0 +1,177 @@
+package org.apache.mesos.executor;
+
+import org.apache.mesos.Protos;
+import org.apache.mesos.offer.TaskUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Each HealthCheckHandler is responsible for executing a single HealthCheck defined for a TaskInfo.
+ * The start() method returns a Future which can be waited upon.  It will only complete if the
+ * HealthCheck has failed, which is another way of saying that it has reached its maximum consecutive
+ * failures limit. Health checks are not run during the grace period as their result would be ignored
+ * in any case.
+ */
+public class HealthCheckHandler {
+    private final Protos.TaskInfo taskInfo;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final HealthCheckStats healthCheckStats;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public static HealthCheckHandler create(
+            Protos.TaskInfo taskInfo,
+            ScheduledExecutorService scheduledExecutorService,
+            HealthCheckStats healthCheckStats)
+            throws HealthCheckValidationException {
+        return new HealthCheckHandler(
+                taskInfo,
+                scheduledExecutorService,
+                healthCheckStats);
+    }
+
+    private HealthCheckHandler(
+            Protos.TaskInfo taskInfo,
+            ScheduledExecutorService scheduledExecutorService,
+            HealthCheckStats healthCheckStats)
+            throws HealthCheckValidationException {
+        validateTask(taskInfo);
+        this.taskInfo = taskInfo;
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.healthCheckStats = healthCheckStats;
+    }
+
+    public ScheduledFuture<?> start() {
+        double interval = taskInfo.getHealthCheck().getIntervalSeconds();
+        double delay = taskInfo.getHealthCheck().getDelaySeconds() +
+                taskInfo.getHealthCheck().getGracePeriodSeconds();
+
+        long intervalMs = (long) (interval * 1000);
+        long delayMs = (long) (delay * 1000);
+
+        logger.info("Scheduling health check.");
+        return scheduledExecutorService.scheduleAtFixedRate(
+                new HealthCheckRunner(taskInfo.getHealthCheck(), healthCheckStats),
+                delayMs,
+                intervalMs,
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void validateTask(Protos.TaskInfo taskInfo) throws HealthCheckValidationException {
+        // Validate TaskInfo
+        if (!taskInfo.hasHealthCheck()) {
+            throw new HealthCheckValidationException("The following task does not contain a HealthCheck: " + taskInfo);
+        }
+
+        // Validate HealthCheck
```

Comment:
maybe add a comment mentioning that this healthcheck configuration is only viewed by us in practice.
per conversation, its never visible to marathon and mesos doesn't support healthchecks on custom executors